src/Stratis.Bitcoin/Utilities/AsyncManualResetEvent.cs

```diff
@@ -0,0 +1,273 @@
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+https://github.com/StephenCleary/AsyncEx
+ */
+
+using System;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    /// <summary>
+    /// An async-compatible manual-reset event.
+    /// </summary>
+    public sealed class AsyncManualResetEvent
+    {
+        /// <summary>
+        /// The object used for synchronization.
+        /// </summary>
+        private readonly object mutex;
+
+        /// <summary>
+        /// The current state of the event.
+        /// </summary>
+        private TaskCompletionSource<object> tcs;
+
+        /// <summary>
+        /// Creates an async-compatible manual-reset event.
+        /// </summary>
+        /// <param name="set">Whether the manual-reset event is initially set or unset.</param>
+        public AsyncManualResetEvent(bool set)
+        {
+            this.mutex = new object();
+            this.tcs = CreateAsyncTaskSource<object>();
+            if (set)
+                this.tcs.TrySetResult(null);
+        }
+
+        /// <summary>
+        /// Creates an async-compatible manual-reset event that is initially unset.
+        /// </summary>
+        public AsyncManualResetEvent()
+            : this(false)
+        {
+        }
+
+        /// <summary>
+        /// Whether this event is currently set. This member is seldom used; code using this member has a high possibility of race conditions.
+        /// </summary>
+        public bool IsSet
+        {
+            get
+            {
+                lock (this.mutex)
+                {
+                    return this.tcs.Task.IsCompleted;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously waits for this event to be set.
+        /// </summary>
+        public Task WaitAsync()
+        {
+            lock (this.mutex)
+            {
+                return this.tcs.Task;
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously waits for this event to be set or for the wait to be canceled.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token used to cancel the wait. If this token is already canceled, this method will first check whether the event is set.</param>
+        public Task WaitAsync(CancellationToken cancellationToken)
+        {
+            Task waitTask = this.WaitAsync();
+            if (waitTask.IsCompleted)
+                return waitTask;
+
+            return WaitAsync(waitTask, cancellationToken);
+        }
+
+        /// <summary>
```

Comment:
i think we don't need this method,
 if we want to wait synchrounously, we will do what we do elsewhere - .GetAwaiter().GetResult()

Reply:
Ok. Wanted to keep it so we can use AsyncManualResetEvent as a normal ManualResetEventSlim but you're right, most likely this methods will never be used.